Assignment no 2.Implement all Prompt Engineering approaches- Interview Approach, COT, TOT. Compare
and contrast all and analyse their applications.  Implement Zero shot and Few shot prompting
and compare their results.

**Install Libraries**

In [1]:
!pip install transformers torch pandas tqdm

**Import Libraries**

In [2]:
import pandas as pd
from transformers import pipeline
from tqdm import tqdm


**Load HuggingFace Mode**l

In [3]:
generator = pipeline(
    "text-generation",
    model="google/flan-t5-large"
)

/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


config.json:   0%|          | 0.00/662 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/3.13G [00:00<?, ?B/s]

Loading weights:   0%|          | 0/558 [00:00<?, ?it/s]

The tied weights mapping and config for this model specifies to tie shared.weight to lm_head.weight, but both are present in the checkpoints, so we will NOT tie them. You should update the config with `tie_word_embeddings=False` to silence this warning


generation_config.json:   0%|          | 0.00/147 [00:00<?, ?B/s]

tokenizer_config.json: 0.00B [00:00, ?B/s]

spiece.model:   0%|          | 0.00/792k [00:00<?, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

special_tokens_map.json: 0.00B [00:00, ?B/s]

The model 'T5ForConditionalGeneration' is not supported for text-generation. Supported models are ['PeftModelForCausalLM', 'AfmoeForCausalLM', 'ApertusForCausalLM', 'ArceeForCausalLM', 'AriaTextForCausalLM', 'BambaForCausalLM', 'BartForCausalLM', 'BertLMHeadModel', 'BertGenerationDecoder', 'BigBirdForCausalLM', 'BigBirdPegasusForCausalLM', 'BioGptForCausalLM', 'BitNetForCausalLM', 'BlenderbotForCausalLM', 'BlenderbotSmallForCausalLM', 'BloomForCausalLM', 'BltForCausalLM', 'CamembertForCausalLM', 'LlamaForCausalLM', 'CodeGenForCausalLM', 'CohereForCausalLM', 'Cohere2ForCausalLM', 'CpmAntForCausalLM', 'CTRLLMHeadModel', 'CwmForCausalLM', 'Data2VecTextForCausalLM', 'DbrxForCausalLM', 'DeepseekV2ForCausalLM', 'DeepseekV3ForCausalLM', 'DiffLlamaForCausalLM', 'DogeForCausalLM', 'Dots1ForCausalLM', 'ElectraForCausalLM', 'Emu3ForCausalLM', 'ErnieForCausalLM', 'Ernie4_5ForCausalLM', 'Ernie4_5_MoeForCausalLM', 'Exaone4ForCausalLM', 'FalconForCausalLM', 'FalconH1ForCausalLM', 'FalconMambaForCausa

**Load dataset**

In [4]:
df = pd.read_csv("dataset.csv")

print(df.head())


   id                                           question  answer
0   1  If a train travels 60 km per hour, how far wil...     300
1   2                                What is 30% of 250?      75
2   3  If 5 workers complete a job in 10 days, how ma...       5
3   4  A shop gives 20% discount on ₹500. What is fin...     400
4   5  If 12 pencils cost ₹60, what is cost of 1 pencil?       5


**Zero Shot**

In [5]:
def zero_shot(q):
    return f"""
Solve this problem.

Question: {q}
Answer:
"""


**Few Shot**

In [6]:
def few_shot(q):
    return f"""
Example:
Q: What is 10% of 200?
A: 20

Example:
Q: If speed is 60 km/h for 2 hours?
A: 120

Now solve:
Q: {q}
A:
"""


**COT(Chain of thoughts)**

In [7]:
def cot_prompt(q):
    return f"""
Question: {q}

Let's think step by step.
Give final answer at end.
"""


**Interview Prompt**

In [8]:
def interview_prompt(q):
    return f"""
Question: {q}

1. What is question asking?
2. What formula or logic is needed?
3. Solve step by step.
4. Give final answer.
"""


**ToT(Tree of Thoughts)**

In [9]:
def tot_prompt(q):
    return f"""
Question: {q}

Give 3 different solution approaches.
Evaluate each approach.
Choose best approach and give final answer.
"""


**Model Run Function**

In [10]:
def run_model(prompt):
    output = generator(prompt, max_length=200)
    return output[0]["generated_text"]


**Run All Prompt Methods On Dataset**

In [11]:
results = []

for _, row in tqdm(df.iterrows(), total=len(df)):

    q = row["question"]
    true_ans = row["answer"]

    res = {
        "question": q,
        "true_answer": true_ans,
        "zero_shot": run_model(zero_shot(q)),
        "few_shot": run_model(few_shot(q)),
        "cot": run_model(cot_prompt(q)),
        "interview": run_model(interview_prompt(q)),
        "tot": run_model(tot_prompt(q))
    }

    results.append(res)


  0%|          | 0/30 [00:00<?, ?it/s]Passing `generation_config` together with generation-related arguments=({'max_length'}) is deprecated and will be removed in future versions. Please pass either a `generation_config` object OR all generation parameters explicitly, but not both.
Both `max_new_tokens` (=256) and `max_length`(=200) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)
Both `max_new_tokens` (=256) and `max_length`(=200) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)
Both `max_new_tokens` (=256) and `max_length`(=200) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/

**Convert Results To DataFrame**

In [12]:
results_df = pd.DataFrame(results)

print(results_df.head())


                                            question  true_answer  \
0  If a train travels 60 km per hour, how far wil...          300   
1                                What is 30% of 250?           75   
2  If 5 workers complete a job in 10 days, how ma...            5   
3  A shop gives 20% discount on ₹500. What is fin...          400   
4  If 12 pencils cost ₹60, what is cost of 1 pencil?            5   

                                           zero_shot  \
0  \nSolve this problem.\n\nQuestion: If a train ...   
1  \nSolve this problem.\n\nQuestion: What is 30%...   
2  \nSolve this problem.\n\nQuestion: If 5 worker...   
3  \nSolve this problem.\n\nQuestion: A shop give...   
4  \nSolve this problem.\n\nQuestion: If 12 penci...   

                                            few_shot  \
0  \nExample:\nQ: What is 10% of 200?\nA: 20\n\nE...   
1  \nExample:\nQ: What is 10% of 200?\nA: 20\n\nE...   
2  \nExample:\nQ: What is 10% of 200?\nA: 20\n\nE...   
3  \nExample:\nQ: What i

**Saved result**

In [13]:
results_df.to_csv("prompt_engineering_results.csv", index=False)

print("Results saved successfully!")


Results saved successfully!


**Accuracy Checking**

In [14]:
def check_correct(pred, true):
    return str(true) in str(pred)

for col in ["zero_shot","few_shot","cot","interview","tot"]:
    results_df[col+"_correct"] = results_df.apply(
        lambda x: check_correct(x[col], x["true_answer"]), axis=1
    )

print(results_df.mean(numeric_only=True))


true_answer          96.100000
zero_shot_correct     0.400000
few_shot_correct      0.466667
cot_correct           0.533333
interview_correct     0.500000
tot_correct           0.400000
dtype: float64


**Experimental results show that Chain-of-Thought prompting achieved the highest accuracy (53%), followed by Interview prompting (50%). Few-shot prompting showed moderate improvement over zero-shot prompting. Tree-of-Thought prompting showed lower performance due to model limitations in evaluating multiple reasoning branches. These results indicate that structured reasoning prompts significantly improve performance in reasoning tasks.**